### PR TITLE
Many-to-many relationships

### PR DESCRIPTION
**Pull request by andrewhayward**
This provides support for many-to-many relationships by allowing `hasMany` relationships to utilise a `via` property.

That is, given the following tables:

``` js
db.table('first', {
  fields: ['id', ...]
  relationships: {
    things: {
      type: 'hasMany',
      local: 'id',
      foreign: { table: 'second', key: 'id' }
    }
  }
})

db.table('second', {
  fields: ['id', ...],
  relationships: {
    things: {
      type: 'hasMany',
      local: 'id',
      foreign: { table: 'first', key: 'id' }
    }
  }
})
```

we can introduce a third table, and add `via` to our relationships to go through that table instead:

``` js
db.table('via', {
  fields: ['first_id', 'second_id']
})

db.table('first', {
  fields: ['id', ...]
  relationships: {
    things: {
      type: 'hasMany',
      local: 'id',
      foreign: { table: 'second', key: 'id' },
      via: { table: 'via', local: 'first_id', foreign: 'second_id' }
    }
  }
})
```

This results in a `first.get(...)` request having a `things` property with many `second` items.

### CODE REVIEW

**Comment by brianloveswords** [comment]
Can you do the `sqlite` tests and docs? Also, we still need docs for nested relationships. DOUBLE ALSO, I've gotta take the time one of these days to fix the test situation so we don't have to duplicate tests for sqlite.

**Comment by brianloveswords** [comment]
TRIPLE ALSO, god damn travis and that failing test. I've never been able to reproduce it :/

**Comment by andrewhayward** [comment]
Yep, tests and docs forthcoming. Just wanted to get this out ASAP.

I thought we had docs for nested relationships? `relationshipsDepth` is listed under [`options`](/brianloveswords/streamsql/blob/master/README.md#options-1), anyway. Let me know if you want more than that.

**Comment by brianloveswords** [comment]
OH SNAP, so it is! Carry on.

(An aside: -1 as a sentinel value is fine, but we should also support `Infinity` – without seeing the docs, that's what I tried and it didn't work)

**Comment by andrewhayward** [comment]
Oh really? `Infinity` works for me. I think.

**Comment by andrewhayward** [comment]
Updating `get.test.js` to use `Infinity` seems to work fine...

``` js
review.getOne({ id: 1 }, {
  debug: true,
  relationships: true,
  relationshipsDepth: Infinity
}, function (err, review) {
  ...
})
```

What didn't work for you?

**Comment by brianloveswords** [comment]
It didn't actually fulfill all of the relationships

``` js
{relationshipsDepth: 2}
// yields
{ id: 1,
     createdOn: '2014-04-17T15:40:55.000Z',
     jsonUrl: 'http://example.org/two',
     earnerId: 'test-user',
     badgeClassId: 1,
     uid: 'test-badge-one',
     imageUrl: null,
     badgeJSONUrl: 'http://example.org/badge.json',
     evidenceUrl: null,
     issuedOn: null,
     expires: null,
     earner:
      { id: 'test-user',
        createdOn: '2014-04-17T15:40:55.000Z',
        metadata: {},
        badges: [],
        _metadata: [] },
     json:
      { url: 'http://example.org/two',
        data: '{"sup": true}',
        updatedOn: '2014-04-17T15:40:55.000Z',
        assertion: [Object],
        badge: [Object],
        issuer: [Object] },
     badgeClass:
      { id: 1,
        createdOn: '2014-04-17T15:40:55.000Z',
        jsonUrl: 'http://example.org/one',
        issuerOrgId: 1,
        name: 'Example Badge Class',
        description: 'I AM ERROR',
        imageUrl: 'http://example.org/image.png',
        criteriaUrl: 'http://example.org/criteria',
        issuerJSONUrl: 'http://example.org/issuer.json',
        json: [Object],
        issuer: [Object] } } }

// -------------------------------------------

{relationshipsDepth: Infinity}
// missing `json` and `badgeClass`
{ id: 1,
     createdOn: '2014-04-17T15:41:03.000Z',
     jsonUrl: 'http://example.org/two',
     earnerId: 'test-user',
     badgeClassId: 1,
     uid: 'test-badge-one',
     imageUrl: null,
     badgeJSONUrl: 'http://example.org/badge.json',
     evidenceUrl: null,
     issuedOn: null,
     expires: null,
     earner:
      { id: 'test-user',
        createdOn: '2014-04-17T15:41:03.000Z',
        metadata: {},
        badges: [],
        _metadata: [] } } }

// -----------------------------------------
// also interesting, if I use too high a value, I get the same result as Infinity
{relationshipsDepth: 3}

{ id: 1,
     createdOn: '2014-04-17T15:43:36.000Z',
     jsonUrl: 'http://example.org/two',
     earnerId: 'test-user',
     badgeClassId: 1,
     uid: 'test-badge-one',
     imageUrl: null,
     badgeJSONUrl: 'http://example.org/badge.json',
     evidenceUrl: null,
     issuedOn: null,
     expires: null,
     earner:
      { id: 'test-user',
        createdOn: '2014-04-17T15:43:36.000Z',
        metadata: {},
        badges: [],
        _metadata: [] } } }
```

**Comment by andrewhayward** [comment]
'Too high a value' and '`Infinity`' are probably hitting the same stumbling block. You'll no doubt get the same issue using `-1` as well. Sounds like a new issue needs to be filed - we're getting a little off-topic!

**Comment by brianloveswords** [comment]
FILING NOW

**Comment by andrewhayward** [comment]
@brianloveswords - docs and sqlite tests pushed. Ready for review.

**Comment by andrewhayward** [comment]
Also, fuck Travis.

File: README.md [hunk 1]
```diff
@@ -204,36 +224,58 @@ const album = db.table('album', {
...
+    bands: {
+      type: 'hasMany',
+      local: 'id',
+      foreign: { table: 'band', key: 'id' },
+      via: { table: 'bandMember', local: 'memberId', foreign: 'bandId' }
+    }
+  }
+})
+
+const band_member = db.table('band_member', {
```
Comment (brianloveswords) [inline README.md]: This should be camelCased, `const bandMember = db.table('bandMember', {`

**Comment by brianloveswords** [comment]
v0.8.0 published
